src/main/java/org/apache/commons/text/numbers/DoubleFormat.java

```diff
@@ -0,0 +1,736 @@
+
+    /** Construct a new instance.
+     * @param factory function used to construct format instances
+     */
+    DoubleFormat(final Function<Builder, DoubleFunction<String>> factory) {
+        this.factory = factory;
+    }
+
+    /** Return a {@link Builder} for constructing formatter functions for this format type.
+     * @return builder instance
+     */
+    public Builder builder() {
+        return new Builder(factory);
+    }
+
+    /** Class for constructing configured format functions for standard double format types.
+     */
+    public static final class Builder {
+
+        /** Default value for the plain format max decimal exponent. */
+        private static final int DEFAULT_PLAIN_FORMAT_MAX_DECIMAL_EXPONENT = 6;
+
+        /** Default value for the plain format min decimal exponent. */
+        private static final int DEFAULT_PLAIN_FORMAT_MIN_DECIMAL_EXPONENT = -3;
+
+        /** Default decimal digit characters. */
+        private static final String DEFAULT_DECIMAL_DIGITS = "0123456789";
+
+        /** Function used to construct format instances. */
+        private final Function<Builder, DoubleFunction<String>> factory;
+
+        /** Maximum number of significant decimal digits in formatted strings. */
+        private int maxPrecision = 0;
+
+        /** Minimum decimal exponent. */
+        private int minDecimalExponent = Integer.MIN_VALUE;
+
+        /** Max decimal exponent to use with plain formatting with the mixed format type. */
+        private int plainFormatMaxDecimalExponent = DEFAULT_PLAIN_FORMAT_MAX_DECIMAL_EXPONENT;
+
+        /** Min decimal exponent to use with plain formatting with the mixed format type. */
+        private int plainFormatMinDecimalExponent = DEFAULT_PLAIN_FORMAT_MIN_DECIMAL_EXPONENT;
+
+        /** String representing infinity. */
+        private String infinity = "Infinity";
+
+        /** String representing NaN. */
+        private String nan = "NaN";
+
+        /** Flag determining if fraction placeholders should be used. */
+        private boolean fractionPlaceholder = true;
+
+        /** Flag determining if signed zero strings are allowed. */
+        private boolean signedZero = true;
+
+        /** String of digit characters 0-9. */
+        private String digits = DEFAULT_DECIMAL_DIGITS;
+
+        /** Decimal separator character. */
+        private char decimalSeparator = '.';
+
+        /** Character used to separate groups of thousands. */
+        private char groupingSeparator = ',';
+
+        /** If true, thousands groups will be separated by the grouping separator. */
+        private boolean groupThousands = false;
+
+        /** Minus sign character. */
+        private char minusSign = '-';
+
+        /** Exponent separator character. */
+        private String exponentSeparator = "E";
+
+        /** Flag indicating if the exponent value should always be included, even if zero. */
+        private boolean alwaysIncludeExponent = false;
```

Comment:
Ah, I had used

```java
        DoubleFunction<String> fmt = DoubleFormat.PLAIN.builder()
                .maxPrecision(3)
                .alwaysIncludeExponent(true)
                .build();
        String s = fmt.apply((1./1));
        System.out.println(s);
```

Expecting it to be something like `1.0E0`, but instead it was `1.0`. Then I saw a comment in the builder's `alwaysIncludeExponent` method, that only SCIENTIFIC formats (including mixed/engineering I think) supported this.

I don't think we could throw an error for cases where we used the `Builder` to construct a formatter and received no errors. In this case, even though I used `.alwaysIncludeExponent(true)` and the builder did not throw any errors, whoever read the docs knows that this won't be used… but unless the user reads the documentation, I guess the final result could be not necessarily what the user expected…. not sure if there's any simple way to prevent cases like this.

Reply:
I've gone back and forth on this. On one hand, having builder options that only apply to certain formats seems confusing. Users must read the docs in order to ensure that an option will apply to their situation. On the other hand, having separate builder types for each format would make the API considerably more complicated. Plus, with a single builder API, users can separate their configuration from the actual format used. For example,
```java
private static DoubleFunction<String> createFormatter(DoubleFormat formatType) {
    return formatType.builder()
        .maxPrecision(6)
        .alwaysIncludeExponent(true) // may or may not apply
        .build();
}
```
So, I'm still leaning toward the current implementation, where the builder type contains a superset of all options across all format types.

Reply:
Fair enough @darkma773r. The API is easy to understand, and the documentation is great too. So let's go with the current design and see if we get feedback from other devs and/or users.